Local Volume Provisioner: Support Reclaim Policy

cofyc opened the pull request:
**What this PR does / why we need it**:

Support Reclaim Policy. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes https://github.com/kubernetes-incubator/external-storage/issues/771

**Special notes for reviewer**:

`Recycle` policy is deprecated (https://kubernetes.io/docs/concepts/storage/persistent-volumes/#recycle), I suggest not supporting it.

Although we can treat it as `Delete` policy, but for `Recycle` if we want to support we should not delete it.

cofyc commented:
/area area/local-volume

cofyc commented:
/assign @msau42

msau42 commented on local-volume/provisioner/pkg/util/api_util.go:
```diff
@@ -97,6 +101,10 @@ func (u *apiUtil) DeleteJob(jobName string, namespace string) error {
 }
 
+func (u *apiUtil) GetStorageClass(scName string) (*storage_v1.StorageClass, error) {
+	return u.client.StorageV1().StorageClasses().Get(scName, metav1.GetOptions{})
```
The RBAC permissions need to add StorageClass Get permissions

cofyc replied:
The `system:persistent-volume-provisioner` clusterrole [we are using](https://github.com/kubernetes-incubator/external-storage/blob/master/local-volume/helm/provisioner/templates/provisioner-cluster-role-binding.yaml#L13) contains `get` permissions on `storageclasses`, no need to add.

msau42 replied:
Can we use informer to get StorageClass objects? To reduce chance of exhausting API quota.

cofyc replied:
@msau42 Updated in https://github.com/kubernetes-incubator/external-storage/pull/776/commits/346c2e28611b81616f4ad21f36913d7228e00712.

- start StorageClass informer controller in Discoverer, then it can get StorageClass from informer cache
- change `common.RuntimeConfig.Client` to be `kubernetes.Interface`, then we can use `fake.Clientset` for testing
- update `Populator.Start` to be same as `Discoverer.Start` which are almost the same
- prefer using `wait.NeverStop` which is more idiomatic in k8s

msau42 commented on local-volume/provisioner/pkg/populator/populator.go:
```diff
@@ -41,8 +39,8 @@ func NewPopulator(config *common.RuntimeConfig) *Populator {
-func (p *Populator) Start() {
+// Start launches the PV informer and waits for cache to sync, returns error if it fails.
+func (p *Populator) Start(stopCh <-chan struct{}) error {
 	_, controller := kcache.NewInformer(
```
Would using SharedInformers be better here for both PV and StorageClass objects?

cofyc replied:
No, IMO [`SharedInformer`](https://github.com/kubernetes/client-go/blob/v7.0.0/tools/cache/shared_informer.go#L34) is preferred if there are multiple listeners for same cache. For examples, two controllers who both will listen on events of PersistentVolume can use `SharedInformer ` to share data cache. In local volume provisioner, PersistentVolums cache is only consumed Populator,  StorageClasse Cache only by Discoverer.

msau42 replied:
The interface seems simpler as you don't have to start separate controllers

cofyc replied:
Agree. I'll updated to use `SharedInformer`.

cofyc replied:
Updated in https://github.com/kubernetes-incubator/external-storage/pull/776/commits/af8c713a0e7c3bf30eeada9aef449f0baf7cda6a.

I've done some more refactoring which I found necessary:

- Use `informers.SharedInformerFactory` to create SharedInformers for PersistentVolumes, StorageClasses, and Jobs.
  - Note that, for Jobs we are using a customized constructor
- With `informers.SharedInformerFactory`, it's easier to start informers and wait for all informers' cache were synced before running controller logic
  - Previously, `JobController` waits for cache in a separate go routine which is not right, because it is not a standalone controller from my perspective.

cofyc commented:
BTW, `Populator` use `0` as resync period, which I think in theory is right choice, because we don't need to reconcile periodically. But in practice, `list` & `watch` probably lost events (someone is tracking [event lost issues](https://github.com/kubernetes/kubernetes/pull/61067)), should we use a non-zero value as last insurance?

cofyc commented:
However, `scheduler` disabled resync period too, i'd found a old discussion why scheduler does not use resync period here: https://github.com/kubernetes/kubernetes/pull/16840#issuecomment-154051747.

I actually don't know what's the right resync period value should to use in this case. I prefer `0` but am  afraid cache may become stale during long running time. But if using non-zero value, don't know how much time is enough.

msau42 commented:
I think it is not necessary for this to be very performant. Maybe a resync every 5 minutes is fine?

msau42 commented:
This lgtm, can you squash into 2 commits? one for informers, and one for the new functionality?

Commit:
- Use SharedInformers instead of Informers in favor of simpler
interfaces.
- Use `informers.SharedInformerFactory` to create SharedInformers for
PersistentVolumes, and Jobs. Note that you are using a customized
constructor for Jobs.
- Start informers and wait for all informers's cache were synced before
running controller logic.

Commit:
Local Volume Provisioner: Support Reclaim Policy

cofyc commented:
Done.

- https://github.com/kubernetes-incubator/external-storage/pull/776/commits/68957947197dc19f4519938541078a62cbcd84da: use SharedInformers and some refactoring
- https://github.com/kubernetes-incubator/external-storage/pull/776/commits/5133a8996ee855b3b359eb01af33fc4a09a96b3a: support StorageClass ReclaimPolicy

About using non-zero resync period, I'll create another PR for it (probably it's better to add a user config allowing users to override like `kube-controller-mananger`'s `--min-resync-period`).

msau42 commented:
/lgtm
Thanks!

msau42 commented:
/assign @childsb
